File: lib/src/main/java/org/simplepresenter/PresenterViewDelegate.java
package org.simplepresenter;

import android.os.Bundle;
import android.os.Handler;

public class PresenterViewDelegate<P extends Presenter> extends ViewDelegate<P> {

    public static final String PRESENTER_ID = "PRESENTER_ID";

    private boolean saveStateCalled;
    private Handler uiHandler = new Handler();

    public PresenterViewDelegate(PresenterView<P> view) {
        setView(view);
    }

    public void onCreate(Bundle state) {
        if (state == null) {
            setPresenter(getView().createPresenter());
            getPresenter().bind(this);
            PresenterHolder.INSTANCE.registerPresenter(getPresenter());
            setResumed(true);
            getPresenter().onViewCreated();
        } else {
            int presenterId = state.getInt(PRESENTER_ID);
            //noinspection unchecked
            setPresenter((P) PresenterHolder.INSTANCE.findPresenterById(presenterId));
            setRestoring(true);
            getPresenter().bind(this);
        }
    }

    public void onPause() {
        setResumed(false);
    }

    public void onResume() {
        setResumed(true);
        if (isRestoring()) {
            getPresenter().onRestoreView();
        }
    }

    public void onSaveInstanceState(Bundle state) {
        state.putInt(PRESENTER_ID, getPresenter().hashCode());
        saveStateCalled = true;
    }

    public void onDestroy() {
        if (!saveStateCalled) {
            getPresenter().onViewDestroyed();
            PresenterHolder.INSTANCE.removePresenterById(getPresenter().hashCode());
        }
        getPresenter().clearReferenceToView();
        setDead(true);
        setPresenter(null);
        setView(null);
    }

    @Override
    public void delegateCommand(final ViewCommand command) {
        uiHandler.post(new Runnable() {
            @Override
            public void run() {
                getView().dispatchCommand(command);
            }
        });
    }
}

File: lib/src/main/java/org/simplepresenter/PresenterViewDelegate.kt
package org.simplepresenter

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import java.util.concurrent.atomic.AtomicInteger

class PresenterViewDelegate(override var view: PresenterView<*>?) : ViewDelegate() {

    private companion object {
        internal const val PRESENTER_ID = "${BuildConfig.APPLICATION_ID}.PresenterViewDelegate.PRESENTER_ID"
        internal val ID_GENERATOR = AtomicInteger(0)
    }

    private var id: Int = 0
    private val store: PresentersStore
    private val uiHandler = Handler(Looper.getMainLooper())
    private var saveStateCalled: Boolean = false

    init {
        store = CommandEngine.Bridge.currentEngine.presentersStore
    }

    fun onCreate(state: Bundle?) {
        if (state == null) {
            id = ID_GENERATOR.incrementAndGet()
            presenter = view?.createPresenter()
            presenter?.bind(this)
            store.savePresenter(id, presenter as Presenter)
            isResumed = true
            presenter?.onViewCreated()
        } else {
            id = state.getInt(PRESENTER_ID)
            presenter = store.loadPresenter(id)
            isRestoring = true
            presenter?.bind(this)
        }
    }

    fun onPause() {
        isResumed = false
    }

    fun onResume() {
        isResumed = true
        if (isRestoring) {
            presenter?.onRestoreView()
        }
    }

    fun onSaveInstanceState(state: Bundle?) {
        if (state == null) {
            throw IllegalStateException("State is null")
        }
        state.putInt(PRESENTER_ID, id)
        saveStateCalled = true
    }

    fun onDestroy() {
        if (!saveStateCalled) {
            presenter?.onViewDestroyed()
            store.deletePresenter(id)
        }
        dispose()
    }

    private fun dispose() {
        presenter?.clearReferenceToView()
        isDead = true
        presenter = null
        view = null
    }

    override fun delegateCommand(command: ViewCommand) {
        uiHandler.post { view?.dispatch(command) }
    }
}